File: longest_increasing_subsequence.py


def longest_increasing_subsequence(nums):
    #This array is the longest subsequence with each element as the END of the subsequence
    dp = [1]*len(nums) # longest subsequence must be at least 1

    for i in range(len(nums)):
        for j in range(i):
            if nums[i] > nums[j]: #if greater, we're appending to the subsequence that ends at nums[j]
                dp[i] = max(dp[i], dp[j] + 1 )
    print(dp)
    return max(dp)


print(longest_increasing_subsequence([1, 9, 7, 4, 7, 13,2,222]))
            

File: longest_increasing_subsequence.cpp
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp(nums.size(), 1);
        for(int i = 0 ; i < nums.size(); i++)
        {
            for (int j = 0 ; j < i; j++)
            {
                if (nums[i]>nums[j])
                {
                    dp[i] = max(dp[i], dp[j]+1);
                }
            }
        }
        auto max = max_element(dp.begin(), dp.end());
        return *max;
        
    }
};